Adding Boruvka algorithm + tests for Minimum Spanning Trees

FranciscoThiesen opened the pull request:
#### Description of Change
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Description: Adding the Boruvka algorithm for the minimum spanning tree problem. It is the earliest algorithm for this problem, it is beautiful and it belongs in this repo (:

Contributors guide: https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md
-->

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: <!-- Please add a one-line description for developers or pull request viewers -->

FranciscoThiesen commented:
@Panquesito7 @realstealthninja reviews appreciated, thank you!

realstealthninja commented on graph/boruvka.cpp:
```diff
@@ -0,0 +1,191 @@
+ * @param edges Edges of the graph.
+ * @return Minimum spanning tree cost.
+ */
+long long boruvka(
```
```suggestion
int64_t boruvka(
```

realstealthninja commented on graph/boruvka.cpp:
```diff
@@ -0,0 +1,191 @@
+ * @return Minimum spanning tree cost.
+ */
+long long boruvka(
+    int num_nodes, int num_edges,
```
wouldnt it make more sense to make these unsigned since num_nodes/num_edges would never be negative?

FranciscoThiesen replied:
Will change it. Thanks for pointing this out.

FranciscoThiesen replied:
You are absolutely right! Just changed it.

realstealthninja commented on graph/boruvka.cpp:
```diff
@@ -0,0 +1,191 @@
+    edges.push_back({1, {0, 1}});
+    edges.push_back({2, {0, 2}});
+    edges.push_back({3, {1, 2}});
+    edges.push_back({4, {1, 3}});
```
as mentioned above if you really want to use int then might want to add tests for cases when some of the values are negative

FranciscoThiesen commented:
@realstealthninja I've addressed your suggestions. Also changed from int -> uint32_t in the union find, where negative values also don't make sense.

Let me know if there are any other issues.

realstealthninja commented on graph/boruvka.cpp:
```diff
@@ -0,0 +1,191 @@
+}
+
+int main() {
+    tests();
```
```suggestion
    tests(); // run self-test implementations
```

realstealthninja commented on graph/boruvka.cpp:
```diff
@@ -0,0 +1,191 @@
+ * https://en.wikipedia.org/wiki/Bor%C5%AFvka%27s_algorithm
+ */
+
+#include <cassert>
+#include <iostream>
+#include <numeric>
+#include <vector>
```
The purpose of these headers must be documented
A simple comment would suffice.

```cpp
#include <header> /// for <reason>
```

FranciscoThiesen replied:
Done!

realstealthninja reviewed (changes requested):
Thank you for your contributions @Panquesito7 could you have a look as well?

FranciscoThiesen commented:
@realstealthninja iterated on the last round of reviews!

FranciscoThiesen commented:
@realstealthninja can you approve the workflows?

Thank you,
Francisco

FranciscoThiesen commented:
@realstealthninja @Panquesito7 can you please approve the required workflows? I've addressed all the comments/suggested changes.

Thank you,
Francisco

FranciscoThiesen commented:
@realstealthninja @Panquesito7

FranciscoThiesen commented:
@realstealthninja @Panquesito7 any updates? I'd like to merge this if possible

FranciscoThiesen commented:
@realstealthninja @Panquesito7

FranciscoThiesen commented:
@realstealthninja @Panquesito7 can you kindly take a look? I have addressed all the changes requested and this is ready to merge. I can't initiate the checks and I am not currently a contributor.
